internal/certrotation/rotation_test.go

```diff
@@ -0,0 +1,335 @@
+	var (
+		now        = time.Now()
+		nowFn      = func() time.Time { return now }
+		nowCA, err = newTestCACertificate(pkix.Name{CommonName: "creator-tests"}, nowFn)
+	)
+	require.NoError(t, err)
+
+	c := certificateRotation{
+		UserInfo:  defaultUserInfo,
+		Hostnames: []string{"example.org"},
+	}
+	cert, err := c.NewCertificate(nowCA, 1*time.Hour)
+	require.NoError(t, err)
+
+	require.Contains(t, cert.Certs[0].ExtKeyUsage, x509.ExtKeyUsageServerAuth)
+	require.Contains(t, cert.Certs[0].ExtKeyUsage, x509.ExtKeyUsageClientAuth)
+	require.Equal(t, defaultUserInfo.GetName(), cert.Certs[0].Subject.CommonName)
+	require.Equal(t, defaultUserInfo.GetUID(), cert.Certs[0].Subject.SerialNumber)
+	require.Equal(t, defaultUserInfo.GetGroups(), cert.Certs[0].Subject.Organization)
+}
+
+func TestCertificateRotation_SetAnnotations(t *testing.T) {
+	var (
+		now        = time.Now()
+		nowFn      = func() time.Time { return now }
+		nowCA, err = newTestCACertificate(pkix.Name{CommonName: "creator-tests"}, nowFn)
+	)
+	require.NoError(t, err)
+
+	c := certificateRotation{Hostnames: []string{"example.org"}}
+
+	annotations := map[string]string{}
+	c.SetAnnotations(nowCA.Config, annotations)
+
+	require.Len(t, annotations, 4)
+	require.Contains(t, annotations, CertificateIssuer)
+	require.Contains(t, annotations, CertificateNotBeforeAnnotation)
+	require.Contains(t, annotations, CertificateNotAfterAnnotation)
+	require.Contains(t, annotations, CertificateHostnames)
+}
+
+func TestCertificateRotation_NeedNewCertificate(t *testing.T) {
+	var (
+		now                 = time.Now()
+		nowFn               = func() time.Time { return now }
+		invalidNotAfter, _  = time.Parse(time.RFC3339, "")
+		invalidNotBefore, _ = time.Parse(time.RFC3339, "")
+		nowCA, _            = newTestCACertificate(pkix.Name{CommonName: "creator-tests"}, nowFn)
+
+		twentyMinutesBeforeNow   = time.Now().Add(-20 * time.Minute)
+		twentyMinutesBeforeNowFn = func() time.Time { return twentyMinutesBeforeNow }
+		twentyMinutesBeforeCA, _ = newTestCACertificate(pkix.Name{CommonName: "creator-tests"}, twentyMinutesBeforeNowFn)
+	)
+
+	tt := []struct {
+		signerFn    func() (*crypto.CA, error)
+		annotations map[string]string
+		desc        string
+		wantReason  string
+		refresh     time.Duration
+	}{
+		{
+			desc: "already expired",
+			annotations: map[string]string{
+				CertificateIssuer:              "creator-tests",
+				CertificateNotAfterAnnotation:  invalidNotAfter.Format(time.RFC3339),
+				CertificateNotBeforeAnnotation: invalidNotBefore.Format(time.RFC3339),
+			},
+			signerFn: func() (*crypto.CA, error) {
+				return nowCA, nil
+			},
+			refresh:    2 * time.Minute,
+			wantReason: "already expired",
+		},
+		{
+			desc: "refresh only when expired",
+			annotations: map[string]string{
+				CertificateIssuer:              "creator-tests",
+				CertificateNotAfterAnnotation:  now.Add(45 * time.Minute).Format(time.RFC3339),
```

Comment:
Is there any reason for `45`?

Reply:
No reason, is just a test, needs to put a number.